### PR TITLE
feat: add internal/fetch package with SSRF-protected HTTP client

### PR DESCRIPTION
**Pull request by ggallen**
## Summary

- Adds `internal/fetch` package implementing an SSRF-hardened HTTP fetcher (ADR-0038 PR 2)
- Enforces HTTPS-only, domain allowlist, DNS pre-resolution with internal IP validation, transport-level IP pinning (DNS rebinding protection), redirect blocking, response size limits, double-encoding rejection, and offline mode
- Includes comprehensive tests covering all security controls and edge cases (22 test cases across 3 test functions)

## Design

- `FetchPolicy` struct controls allowed domains, max size, timeout, and offline mode
- `FetchURL` performs the full SSRF-safe fetch pipeline: offline check, HTTPS-only, `%25` rejection, domain allowlist, DNS resolution, IP validation, pinned dial, no redirects, 200-only, size limit
- `isInternalIP` normalizes IPv4-mapped IPv6 and checks loopback/private/link-local/unspecified/multicast plus CIDRs for 0.0.0.0/8, 100.64.0.0/10, 198.18.0.0/15
- `ComputeSHA256` utility for content integrity verification
- Only stdlib imports

## Test plan

- [x] `TestFetchURL` subtests: HTTPSOnly, DomainAllowlist, WildcardDomain, NoRedirects, SizeLimit, Timeout, OfflineMode, DoubleEncoding, NonOKStatus, Success
- [x] `TestIsInternalIP`: loopback, RFC1918, link-local, CGNAT, benchmark, unspecified, multicast, IPv4-mapped IPv6, public IPs (22 cases)
- [x] `TestComputeSHA256`: known-value verification
- [x] `go test -race -vet=all ./internal/fetch/...` passes
- [x] All pre-commit hooks pass (go vet, gitleaks, trailing whitespace, etc.)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/fetch/fetch.go [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+	// Determine the port (default 443 for HTTPS).
+	port := parsed.Port()
+	if port == "" {
+		port = "443"
+	}
+
+	// 7. Build transport with DNS-rebinding protection.
+	transport := &http.Transport{
+		DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/fetch/fetch.go]: Fixed — replaced `net.DialTimeout` with `(&net.Dialer{Timeout: policy.Timeout}).DialContext(dialCtx, ...)` so cancelled contexts interrupt in-flight dials immediately. Pushed in 5fad560a.

File: internal/fetch/fetch.go [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+	}
+
+	// Pick the first address for connection pinning.
+	pinnedIP := addrs[0].IP.String()
+
+	// Determine the port (default 443 for HTTPS).
+	port := parsed.Port()
+	if port == "" {
+		port = "443"
+	}
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/fetch/fetch.go:123]: Fixed — DialContext now iterates all pre-validated IPs in order, attempting each for IPv4/IPv6 fallback instead of pinning to only the first. Pushed in 5fad560a.

File: internal/fetch/fetch.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+	if int64(len(data)) > policy.MaxSizeBytes {
+		return nil, errTooLarge
+	}
+
+	return data, nil
+}
+
+// isAllowedDomain checks whether hostname matches any entry in the allowed
+// list. Entries may be exact hostnames or wildcard patterns like "*.example.com"
+// which match any subdomain of example.com (but not example.com itself).
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/fetch/fetch.go:196]: Fixed — added RFC 5737 documentation/TEST-NET CIDRs (192.0.2.0/24, 198.51.100.0/24, 203.0.113.0/24) to `internalCIDRs` in `init()`, matching the ranges already blocked by `internal/security/ssrf.go`. Added test cases for all three ranges. Pushed in 5fad560a.

**Review by rh-hemartin — APPROVED** [review]

File: internal/fetch/fetch.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+		// 8. Block redirects.
+		CheckRedirect: func(_ *http.Request, _ []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, rawURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("fetch: failed to create request: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/fetch/fetch.go:168]: Fixed — added `pattern = strings.ToLower(pattern)` at the top of the `isAllowedDomain` loop so mixed-case entries work correctly.

File: internal/fetch/fetch_test.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+		Timeout:        5 * time.Second,
+		tlsConfig:      srv.TLS.Clone(),
+		skipIPCheck:    true,
+	}
+	// Ensure the TLS config can verify the test server's certificate
+	// when connecting via IP address.
+	policy.tlsConfig.InsecureSkipVerify = true
+
+	return srv, policy
+}
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/fetch/fetch_test.go:46]: Fixed — updated comment to accurately describe the `InsecureSkipVerify` behavior.

File: internal/fetch/fetch.go [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+	// Determine the port (default 443 for HTTPS).
+	port := parsed.Port()
+	if port == "" {
+		port = "443"
+	}
+
+	// 6b. Port restriction.
+	allowedPorts := policy.AllowedPorts
+	if len(allowedPorts) == 0 {
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/fetch/fetch.go:127]: Fixed — added `defer transport.CloseIdleConnections()` after creating the transport.
